### ISSUE TITLE
Update keepass.install to check pre-existing run state.

### ISSUE THREAD

**Issue by bcurran3** [pauby/ChocoPackages#13]
KeePass always fails on upgrades for me since I always have KeePass running and it's native installer doesn't currently support stopping KeePass before upgrading it.

I'd like to request you add that feature to the install script. 

The below code added to chocolateyinstall should do the trick:

if ((get-process "KeePass" -ea SilentlyContinue) -eq $Null){ 
$KeePassRunning="False" # Always false on new install.
} else { 
$KeePassRunning="True"
Write-Host "KeePass found running." -foreground magenta 
}

....current script here.....

if ($KeePassRunning -eq "True"){
$bits = Get-ProcessorBits
Write-Host "Running KeePass" -foreground magenta 
if ($bits -eq 64){
& "${env:PROGRAMFILES(x86)}\KeePass Password Safe 2\KeePass.exe"
} else {
& "$env:PROGRAMFILES\KeePass Password Safe 2\KeePass.exe"
}
}

**Comment by bcurran3** [comment]
UPDATE: I made a helper to make this process even easier:
https://chocolatey.org/packages/chocolatey-misc-helpers.extension/0.0.1

* Checks for and stops $ProcessName if running
* $ProcessWasRunning can be checked later in scripts to re-run the process if it was running previously before stopped
* $ProcessFullPath is the full path and executable of the stopped process
* i.e. to re-run the process after an upgrade - if ($ProcessWasRunning -eq "True") {$ProcessFullPath}
 
function Start-CheckandStop($ProcessName){
$ProcessWasRunning="False"
 
if((Get-Process $ProcessName -ea SilentlyContinue) -eq $Null){ 
    Write-Host "$ProcessName currently NOT running." -foreground green
  }else{ 
    Write-Host "Stopping $ProcessName process..." -foreground yellow
    Stop-Process -ProcessName "$ProcessName" -Force
  $ProcessWasRunning = "True"
  $ProcessFullPath = (Get-Process $ProcessName).path
  }
}`

**Comment by bcurran3** [comment]
bump

**Comment by pauby** [comment]
@bcurran3 Thanks for submitting this and apologies for not coming back. 

Nobody has highlighted this as an issue to me. I'm not keen on killing the KeePass process because the user could have unsaved items which, after I just tested it, will not be saved.

For that reason I'm going to leave the installer as is. Like many Chocolatey packages (or upgrading software in general), you should close it before upgrading.

**Comment by jtcmedia** [comment]
I'm with @bcurran3 on this. I have over 100 packages installed with chocolatey and to go through and manually check which ones are running when upgrading seems silly. There is an option in KeePass to automatically save on close of database and/or automatically save after modifying an entry 
![image](https://user-images.githubusercontent.com/3260027/45531994-108e9000-b7a7-11e8-9071-125daacdb39f.png)

I currently terminate a running process of a pkg that I maintain and haven't had any complaints. Hope you reconsider this.

**Comment by pauby** [comment]
@jtcmedia I appreciate the problem on this. When I tested this Keepass does not shut down gracefully. It doesn't save any passwords that you may not have saved so could result in a loss of data which is not acceptable.

How would you propose to square that circle?

**Comment by bcurran3** [comment]
Thanks @jtcmedia for speaking up.

Well, here's how you do it...

This is the COMPROMISE METHOD:

add to chocolatyInstall.ps1
check if KeePass task is running
if yes
open and parse $env:AppData\KeePass\KeePass.config.xml
search for <AutoSaveAfterEntryEdit>
if true found, kill KeePass task and continue upgrade
if false found throw as the upgrade won't work anyway under the condition of KeePass running and you protect any possible unsaved user data

if the KeePass task is not running all this is moot and continue running the upgrade 

Where there's a Will(iam), there's a way!

**Comment by pauby** [comment]
Does that cover the use case of where the entry box is still open, has been modified and OK hasn't been clicked?

**Comment by bcurran3** [comment]
> Does that cover the use case of where the entry box is still open, has been modified and OK hasn't been clicked?

Probably not.

Just how often do you think an end user will be editing an entry box while trying to do an upgrade of the package? 

Now compare that to the number of upgrade failures due the the fact that KeePass won't upgrade if it's running?

Which one is the more realistic and repetitive scenario?

I don't like package failures. If I tell a package to upgrade, I expect it's going to upgrade.

**Comment by bcurran3** [comment]
In the event you decide to fight closing the program before installation until your dying breath...
Upon failure write a warning to the end user that the package failed due to the KeePass already running and to close the program and try again.

**Comment by bcurran3** [comment]
I see we're going to have to agree to disagree.

Another problem is that the keepass meta package WILL upgrade and keepass.install won't. I think a warning to quit and try again is a reasonable action. My keepass package is 2.40. My KeePass program is stuck at 2.39.1 until I remember (now) to quit it and run the update again. (Yes, I understand this not a package problem.)

**Comment by pauby** [comment]
> I think a warning to quit and try again is a reasonable action

I agree is not unreasonable. As I said Chocolatey will alert the user to it. You are welcome to submit a PR for that (without adding additional dependencies) or I will look at it when I get some time.

**Comment by jtcmedia** [comment]
The upgrade breaks keepass.install. When I do a `choco list -l` after, it doesn't show up. I have to re-install keepass.install to get the upgrade. At least a warning would prevent this problem. Your pkg isn't alone. Git has the same issue if you have a git bash shell open.

> Another problem is that the keepass meta package WILL upgrade and keepass.install won't. I think a warning to quit and try again is a reasonable action. My keepass package is 2.40. My KeePass program is stuck at 2.39.1 until I remember (now) to quit it and run the update again. (Yes, I understand this not a package problem.)

**Comment by pauby** [comment]
@stefan-muc See [here](https://github.com/pauby/ChocoPackages/issues/13#issuecomment-422324780) - PR's welcome for that.

**Comment by pauby** [comment]
Does that pop up the Keypass window asking the user to save or does it do it in the background?

**Comment by stefan-muc** [comment]
veto and petition to re-open
